cachi2/core/package_managers/gomod.py

```diff
@@ -323,9 +327,10 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[GoPackage]:
         cmd = [*go_list, "-deps", "-json=ImportPath,Module,Standard,Deps", pattern]
         return map(GoPackage.parse_obj, load_json_stream(_run_gomod_cmd(cmd, run_params)))
 
-    package_modules = [mod for pkg in go_list_deps("all") if (mod := pkg.module) and not mod.main]
+    package_modules = [module for pkg in go_list_deps("all") if (module := pkg.module) and not module.main]
 
-    all_modules = list(set(chain(downloaded_modules, package_modules)))
+    # package_modules have the replace data, so they should take precedence in the deduplication
+    all_modules = _deduplicate_resolved_modules(chain(package_modules, downloaded_modules))
```

Comment:
Since the order really does matter here, maybe pass it as two separate arguments? So that `_deduplicate_resolved_modules(download_modules, package_modules)` is in charge of handling them in the correct order, rather than having this be a responsibility of the caller